File: packages/philjs-rocket/src/middleware.ts
/**
 * PhilJS Rocket Middleware
 *
 * Middleware components for Rocket framework integration.
 * These are implemented as fairings in Rocket but exposed as
 * middleware-like abstractions for TypeScript configuration.
 */

import type {
  RocketConfig,
  RocketSSRConfig,
  RocketCORSConfig,
  RocketSecurityConfig,
  FairingContext,
  FairingResponse,
} from './types';

// ============================================================================
// SSR Middleware
// ============================================================================

/**
 * SSR middleware configuration
 */
export interface SSRMiddlewareConfig {
  /** Enable streaming SSR */
  streaming?: boolean;
  /** Inject hydration scripts */
  hydration?: boolean;
  /** Cache rendered pages */
  cache?: boolean;
  /** Cache TTL in seconds */
  cacheTTL?: number;
  /** Routes to apply SSR to (glob patterns) */
  routes?: string[];
  /** Routes to exclude from SSR */
  excludeRoutes?: string[];
}

/**
 * Create SSR middleware configuration
 */
export function createSSRMiddleware(config: SSRMiddlewareConfig = {}): SSRMiddleware {
  return new SSRMiddleware(config);
}

/**
 * SSR Middleware class
 */
export class SSRMiddleware {
  private config: SSRMiddlewareConfig;

  constructor(config: SSRMiddlewareConfig) {
    this.config = {
      streaming: false,
      hydration: true,
      cache: false,
      cacheTTL: 300,
      routes: ['/**'],
      excludeRoutes: ['/api/**', '/static/**'],
      ...config,
    };
  }

  /**
   * Check if a path should be SSR'd
   */
  shouldSSR(path: string): boolean {
    // Check exclude routes first
    for (const pattern of this.config.excludeRoutes || []) {
      if (this.matchPath(path, pattern)) {
        return false;
      }
    }

    // Check include routes
    for (const pattern of this.config.routes || ['/**']) {
      if (this.matchPath(path, pattern)) {
        return true;
      }
    }

    return false;
  }

  /**
   * Simple glob pattern matching
   */
  private matchPath(path: string, pattern: string): boolean {
    const regexPattern = pattern
      .replace(/\*\*/g, '.*')
      .replace(/\*/g, '[^/]*')
      .replace(/\?/g, '.');
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(path);
  }

  /**
   * Get cache key for a request
   */
  getCacheKey(ctx: FairingContext): string {
    return `ssr:${ctx.method}:${ctx.path}:${JSON.stringify(ctx.query)}`;
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    return `
use philjs_rocket::middleware::SsrMiddleware;

let ssr_middleware = SsrMiddleware::new()
    .streaming(${this.config.streaming})
    .hydration(${this.config.hydration})
    .cache(${this.config.cache})
    .cache_ttl(${this.config.cacheTTL});
`.trim();
  }

  getConfig(): SSRMiddlewareConfig {
    return this.config;
  }
}

// ============================================================================
// CORS Middleware
// ============================================================================

/**
 * CORS middleware configuration
 */
export interface CORSMiddlewareConfig extends RocketCORSConfig {}

/**
 * Create CORS middleware configuration
 */
export function createCORSMiddleware(config: CORSMiddlewareConfig = {}): CORSMiddleware {
  return new CORSMiddleware(config);
}

/**
 * CORS Middleware class
 */
export class CORSMiddleware {
  private config: CORSMiddlewareConfig;

  constructor(config: CORSMiddlewareConfig) {
    this.config = {
      enabled: true,
      origins: ['*'],
      methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
      headers: ['Content-Type', 'Authorization', 'X-Requested-With'],
      credentials: false,
      maxAge: 3600,
      exposeHeaders: [],
      ...config,
    };
  }

  /**
   * Check if an origin is allowed
   */
  isOriginAllowed(origin: string): boolean {
    const origins = this.config.origins || ['*'];
    if (origins.includes('*')) {
      return true;
    }
    return origins.includes(origin);
  }

  /**
   * Get CORS headers for a request
   */
  getCORSHeaders(origin: string): Record<string, string> {
    const headers: Record<string, string> = {};

    if (this.isOriginAllowed(origin)) {
      headers['Access-Control-Allow-Origin'] = this.config.origins?.includes('*') ? '*' : origin;
    }

    if (this.config.credentials) {
      headers['Access-Control-Allow-Credentials'] = 'true';
    }

    if (this.config.methods && this.config.methods.length > 0) {
      headers['Access-Control-Allow-Methods'] = this.config.methods.join(', ');
    }

    if (this.config.headers && this.config.headers.length > 0) {
      headers['Access-Control-Allow-Headers'] = this.config.headers.join(', ');
    }

    if (this.config.exposeHeaders && this.config.exposeHeaders.length > 0) {
      headers['Access-Control-Expose-Headers'] = this.config.exposeHeaders.join(', ');
    }

    if (this.config.maxAge) {
      headers['Access-Control-Max-Age'] = this.config.maxAge.toString();
    }

    return headers;
  }

  /**
   * Handle preflight request
   */
  handlePreflight(ctx: FairingContext): FairingResponse {
    const origin = ctx.headers['origin'] || '*';
    return {
      status: 204,
      headers: this.getCORSHeaders(origin),
    };
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    const originsStr = (this.config.origins || []).map(o => `"${o}"`).join(', ');
    const methodsStr = (this.config.methods || []).map(m => `"${m}"`).join(', ');
    const headersStr = (this.config.headers || []).map(h => `"${h}"`).join(', ');

    return `
use philjs_rocket::middleware::CorsMiddleware;

let cors = CorsMiddleware::new()
    .origins(&[${originsStr}])
    .methods(&[${methodsStr}])
    .headers(&[${headersStr}])
    .credentials(${this.config.credentials})
    .max_age(${this.config.maxAge});
`.trim();
  }

  getConfig(): CORSMiddlewareConfig {
    return this.config;
  }
}

// ============================================================================
// Security Headers Middleware
// ============================================================================

/**
 * Security headers middleware configuration
 */
export interface SecurityMiddlewareConfig extends RocketSecurityConfig {}

/**
 * Create security headers middleware
 */
export function createSecurityMiddleware(config: SecurityMiddlewareConfig = {}): SecurityMiddleware {
  return new SecurityMiddleware(config);
}

/**
 * Security Headers Middleware class
 */
export class SecurityMiddleware {
  private config: SecurityMiddlewareConfig;

  constructor(config: SecurityMiddlewareConfig) {
    this.config = {
      enabled: true,
      csp: "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'",
      frameOptions: 'DENY',
      contentTypeOptions: 'nosniff',
      referrerPolicy: 'strict-origin-when-cross-origin',
      hsts: 'max-age=31536000; includeSubDomains',
      ...config,
    };
  }

  /**
   * Get security headers
   */
  getSecurityHeaders(): Record<string, string> {
    const headers: Record<string, string> = {};

    if (this.config.csp) {
      headers['Content-Security-Policy'] = this.config.csp;
    }

    if (this.config.frameOptions) {
      headers['X-Frame-Options'] = this.config.frameOptions;
    }

    if (this.config.contentTypeOptions) {
      headers['X-Content-Type-Options'] = this.config.contentTypeOptions;
    }

    if (this.config.referrerPolicy) {
      headers['Referrer-Policy'] = this.config.referrerPolicy;
    }

    if (this.config.hsts) {
      headers['Strict-Transport-Security'] = this.config.hsts;
    }

    if (this.config.permissionsPolicy) {
      headers['Permissions-Policy'] = this.config.permissionsPolicy;
    }

    return headers;
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    return `
use philjs_rocket::middleware::SecurityMiddleware;

let security = SecurityMiddleware::new()
    .csp("${this.config.csp}")
    .frame_options("${this.config.frameOptions}")
    .content_type_options("${this.config.contentTypeOptions}")
    .referrer_policy("${this.config.referrerPolicy}")
    .hsts("${this.config.hsts}");
`.trim();
  }

  getConfig(): SecurityMiddlewareConfig {
    return this.config;
  }
}

// ============================================================================
// Compression Middleware
// ============================================================================

/**
 * Compression middleware configuration
 */
export interface CompressionMiddlewareConfig {
  /** Enable compression */
  enabled?: boolean;
  /** Minimum size to compress (bytes) */
  minSize?: number;
  /** Compression level (1-9) */
  level?: number;
  /** Enable gzip */
  gzip?: boolean;
  /** Enable brotli */
  brotli?: boolean;
  /** Content types to compress */
  contentTypes?: string[];
}

/**
 * Create compression middleware
 */
export function createCompressionMiddleware(config: CompressionMiddlewareConfig = {}): CompressionMiddleware {
  return new CompressionMiddleware(config);
}

/**
 * Compression Middleware class
 */
export class CompressionMiddleware {
  private config: CompressionMiddlewareConfig;

  constructor(config: CompressionMiddlewareConfig) {
    this.config = {
      enabled: true,
      minSize: 1024,
      level: 6,
      gzip: true,
      brotli: true,
      contentTypes: [
        'text/html',
        'text/css',
        'text/javascript',
        'application/javascript',
        'application/json',
        'image/svg+xml',
      ],
      ...config,
    };
  }

  /**
   * Check if content type should be compressed
   */
  shouldCompress(contentType: string): boolean {
    const types = this.config.contentTypes || [];
    return types.some(t => contentType.startsWith(t));
  }

  /**
   * Get best compression algorithm based on Accept-Encoding
   */
  getBestEncoding(acceptEncoding: string): 'br' | 'gzip' | null {
    if (this.config.brotli && acceptEncoding.includes('br')) {
      return 'br';
    }
    if (this.config.gzip && acceptEncoding.includes('gzip')) {
      return 'gzip';
    }
    return null;
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    return `
use philjs_rocket::middleware::CompressionMiddleware;

let compression = CompressionMiddleware::new()
    .min_size(${this.config.minSize})
    .level(${this.config.level})
    .gzip(${this.config.gzip})
    .brotli(${this.config.brotli});
`.trim();
  }

  getConfig(): CompressionMiddlewareConfig {
    return this.config;
  }
}

// ============================================================================
// Rate Limiting Middleware
// ============================================================================

/**
 * Rate limiting middleware configuration
 */
export interface RateLimitMiddlewareConfig {
  /** Enable rate limiting */
  enabled?: boolean;
  /** Requests per window */
  limit?: number;
  /** Window size in seconds */
  window?: number;
  /** Key extractor: 'ip' | 'user' | 'apiKey' */
  keyBy?: 'ip' | 'user' | 'apiKey';
  /** Custom key header (for apiKey) */
  keyHeader?: string;
  /** Routes to apply rate limiting */
  routes?: string[];
  /** Routes to exclude */
  excludeRoutes?: string[];
}

/**
 * Create rate limiting middleware
 */
export function createRateLimitMiddleware(config: RateLimitMiddlewareConfig = {}): RateLimitMiddleware {
  return new RateLimitMiddleware(config);
}

/**
 * Rate Limit Middleware class
 */
export class RateLimitMiddleware {
  private config: RateLimitMiddlewareConfig;

  constructor(config: RateLimitMiddlewareConfig) {
    this.config = {
      enabled: true,
      limit: 100,
      window: 60,
      keyBy: 'ip',
      keyHeader: 'X-API-Key',
      routes: ['/api/**'],
      excludeRoutes: [],
      ...config,
    };
  }

  /**
   * Get rate limit key from context
   */
  getKey(ctx: FairingContext): string {
    switch (this.config.keyBy) {
      case 'ip':
        return ctx.remoteAddr || 'unknown';
      case 'user':
        return ctx.headers['authorization'] || 'anonymous';
      case 'apiKey':
        return ctx.headers[this.config.keyHeader?.toLowerCase() || 'x-api-key'] || 'no-key';
      default:
        return ctx.remoteAddr || 'unknown';
    }
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    return `
use philjs_rocket::middleware::RateLimitMiddleware;

let rate_limit = RateLimitMiddleware::new()
    .limit(${this.config.limit})
    .window(${this.config.window})
    .key_by("${this.config.keyBy}");
`.trim();
  }

  getConfig(): RateLimitMiddlewareConfig {
    return this.config;
  }
}

// ============================================================================
// Tracing Middleware
// ============================================================================

/**
 * Tracing middleware configuration
 */
export interface TracingMiddlewareConfig {
  /** Enable tracing */
  enabled?: boolean;
  /** Log level */
  level?: 'trace' | 'debug' | 'info' | 'warn' | 'error';
  /** Include request body */
  logBody?: boolean;
  /** Include response body */
  logResponseBody?: boolean;
  /** Include headers */
  logHeaders?: boolean;
  /** Redact sensitive headers */
  redactHeaders?: string[];
}

/**
 * Create tracing middleware
 */
export function createTracingMiddleware(config: TracingMiddlewareConfig = {}): TracingMiddleware {
  return new TracingMiddleware(config);
}

/**
 * Tracing Middleware class
 */
export class TracingMiddleware {
  private config: TracingMiddlewareConfig;

  constructor(config: TracingMiddlewareConfig) {
    this.config = {
      enabled: true,
      level: 'info',
      logBody: false,
      logResponseBody: false,
      logHeaders: true,
      redactHeaders: ['authorization', 'cookie', 'x-api-key'],
      ...config,
    };
  }

  /**
   * Redact sensitive headers
   */
  redactHeaders(headers: Record<string, string>): Record<string, string> {
    const redacted = { ...headers };
    const toRedact = this.config.redactHeaders || [];

    for (const header of toRedact) {
      if (redacted[header.toLowerCase()]) {
        redacted[header.toLowerCase()] = '[REDACTED]';
      }
    }

    return redacted;
  }

  /**
   * Generate Rust fairing code
   */
  toRustCode(): string {
    return `
use philjs_rocket::middleware::TracingMiddleware;
use tracing::Level;

let tracing = TracingMiddleware::new()
    .level(Level::${this.config.level?.toUpperCase()})
    .log_headers(${this.config.logHeaders})
    .log_body(${this.config.logBody});
`.trim();
  }

  getConfig(): TracingMiddlewareConfig {
    return this.config;
  }
}

File: packages/philjs-rocket/rust/src/middleware.rs
//! Middleware utilities for PhilJS Rocket integration
//!
//! Note: Rocket uses fairings instead of traditional middleware.
//! This module provides middleware-like patterns and helpers.

use rocket::fairing::{Fairing, Info, Kind};
use rocket::http::Header;
use rocket::{Request, Response, Data};
use std::time::Instant;
use tracing::info;

/// Request timing middleware (as fairing)
pub struct TimingMiddleware;

impl TimingMiddleware {
    /// Create a new timing middleware
    pub fn new() -> Self {
        Self
    }
}

impl Default for TimingMiddleware {
    fn default() -> Self {
        Self::new()
    }
}

#[rocket::async_trait]
impl Fairing for TimingMiddleware {
    fn info(&self) -> Info {
        Info {
            name: "Request Timing",
            kind: Kind::Request | Kind::Response,
        }
    }

    async fn on_request(&self, request: &mut Request<'_>, _data: &mut Data<'_>) {
        request.local_cache(|| Instant::now());
    }

    async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
        let start = request.local_cache(|| Instant::now());
        let duration = start.elapsed();

        response.set_header(Header::new(
            "X-Response-Time",
            format!("{:.3}ms", duration.as_secs_f64() * 1000.0),
        ));
    }
}

/// Request ID middleware
pub struct RequestIdMiddleware;

impl RequestIdMiddleware {
    /// Create a new request ID middleware
    pub fn new() -> Self {
        Self
    }
}

impl Default for RequestIdMiddleware {
    fn default() -> Self {
        Self::new()
    }
}

#[rocket::async_trait]
impl Fairing for RequestIdMiddleware {
    fn info(&self) -> Info {
        Info {
            name: "Request ID",
            kind: Kind::Request | Kind::Response,
        }
    }

    async fn on_request(&self, request: &mut Request<'_>, _data: &mut Data<'_>) {
        let request_id = uuid::Uuid::new_v4().to_string();
        request.local_cache(|| request_id);
    }

    async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
        let request_id = request.local_cache(|| uuid::Uuid::new_v4().to_string());
        response.set_header(Header::new("X-Request-Id", request_id.clone()));
    }
}

/// Compression middleware settings
pub struct CompressionConfig {
    /// Minimum size to compress (bytes)
    pub min_size: usize,
    /// Compression level (1-9)
    pub level: u32,
}

impl Default for CompressionConfig {
    fn default() -> Self {
        Self {
            min_size: 1024,
            level: 6,
        }
    }
}

/// Rate limiting configuration
pub struct RateLimitConfig {
    /// Maximum requests per window
    pub max_requests: u32,
    /// Window duration in seconds
    pub window_seconds: u64,
}

impl Default for RateLimitConfig {
    fn default() -> Self {
        Self {
            max_requests: 100,
            window_seconds: 60,
        }
    }
}

/// Logging middleware
pub struct LoggingMiddleware {
    /// Log level
    level: tracing::Level,
    /// Include request body
    log_body: bool,
}

impl LoggingMiddleware {
    /// Create a new logging middleware
    pub fn new() -> Self {
        Self {
            level: tracing::Level::INFO,
            log_body: false,
        }
    }

    /// Set log level
    pub fn level(mut self, level: tracing::Level) -> Self {
        self.level = level;
        self
    }

    /// Enable body logging
    pub fn log_body(mut self, enabled: bool) -> Self {
        self.log_body = enabled;
        self
    }
}

impl Default for LoggingMiddleware {
    fn default() -> Self {
        Self::new()
    }
}

#[rocket::async_trait]
impl Fairing for LoggingMiddleware {
    fn info(&self) -> Info {
        Info {
            name: "Request Logging",
            kind: Kind::Request | Kind::Response,
        }
    }

    async fn on_request(&self, request: &mut Request<'_>, _data: &mut Data<'_>) {
        let method = request.method();
        let uri = request.uri();

        info!(
            method = %method,
            uri = %uri,
            "Incoming request"
        );

        request.local_cache(|| Instant::now());
    }

    async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
        let start = request.local_cache(|| Instant::now());
        let duration = start.elapsed();
        let status = response.status();

        info!(
            status = %status.code,
            duration = ?duration,
            "Request completed"
        );
    }
}

/// Cache control configuration
#[derive(Clone)]
pub struct CacheControl {
    /// Cache directive
    directive: String,
    /// Max age in seconds
    max_age: Option<u32>,
}

impl CacheControl {
    /// Create no-cache directive
    pub fn no_cache() -> Self {
        Self {
            directive: "no-cache, no-store, must-revalidate".to_string(),
            max_age: None,
        }
    }

    /// Create public cache directive
    pub fn public(max_age: u32) -> Self {
        Self {
            directive: format!("public, max-age={}", max_age),
            max_age: Some(max_age),
        }
    }

    /// Create private cache directive
    pub fn private(max_age: u32) -> Self {
        Self {
            directive: format!("private, max-age={}", max_age),
            max_age: Some(max_age),
        }
    }

    /// Create immutable cache directive
    pub fn immutable(max_age: u32) -> Self {
        Self {
            directive: format!("public, max-age={}, immutable", max_age),
            max_age: Some(max_age),
        }
    }

    /// Get the directive string
    pub fn as_str(&self) -> &str {
        &self.directive
    }
}

/// Middleware chain helper
pub struct MiddlewareChain {
    fairings: Vec<Box<dyn Fairing>>,
}

impl MiddlewareChain {
    /// Create a new middleware chain
    pub fn new() -> Self {
        Self {
            fairings: Vec::new(),
        }
    }

    /// Add a fairing to the chain
    pub fn add<F: Fairing + 'static>(mut self, fairing: F) -> Self {
        self.fairings.push(Box::new(fairing));
        self
    }

    /// Get all fairings
    pub fn into_fairings(self) -> Vec<Box<dyn Fairing>> {
        self.fairings
    }
}

impl Default for MiddlewareChain {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cache_control() {
        let no_cache = CacheControl::no_cache();
        assert!(no_cache.as_str().contains("no-cache"));

        let public = CacheControl::public(3600);
        assert!(public.as_str().contains("public"));
        assert!(public.as_str().contains("3600"));

        let immutable = CacheControl::immutable(31536000);
        assert!(immutable.as_str().contains("immutable"));
    }

    #[test]
    fn test_middleware_chain() {
        let chain = MiddlewareChain::new()
            .add(TimingMiddleware::new())
            .add(RequestIdMiddleware::new())
            .add(LoggingMiddleware::new());

        let fairings = chain.into_fairings();
        assert_eq!(fairings.len(), 3);
    }
}
